Upgrade to Hibernate Validator 6.1.0.Final, RESTEasy 4.4.0.CR1 and move to Jakarta artifacts

gsmet opened the pull request:
I moved most of the artifacts to Jakarta but not all as RESTEasy optional dependencies are heavily relying on two JBoss API spec artifacts for JAX-RS and JAXB (but not entirely consistently, I will report that). These 2 artifacts were updated to Jakarta.

Note that I added a lot of exclusions to be sure we don't have any duplicates: we had a few duplications present in the current code base.

I have another branch with a complete move to Jakarta but:
- it adds a lot more exclusions + explicit dependencies as JAX-RS and JAXB are injected in a lot of places
- every time you add an optional RESTEasy dependencies (typically the one for RX-Java2), you end up with duplicate artifacts (or requires exclusions + explicit dependencies)
So I thought this was the best compromise for now.

Draft status for now as we need to discuss this.

@FroMage could you have a look at https://github.com/quarkusio/quarkus/pull/4907/commits/e9af116176410384c00b13114e78d5529dbaa44e#diff-806baf513aab8570dfd629e9eaff1dadR136 , I'm not entirely sure where I should get the remote address from. Usually the header don't provide the really remote one so I was wondering exactly what you had in mind? Note that I might have missed something in Vert.x.

gsmet committed:
Upgrade to Hibernate Validator 6.1.0.Final

gsmet committed:
Update to RESTEasy 4.4.0.CR1

gsmet committed:
Move to Jakarta artifacts

gsmet committed:
Get rid of JBoss API spec artifacts to unify things with Jakarta, except...

for JAXB and JAX-RS as a lot of the optional RESTEasy artifacts depend
on them so you get them in every time you add an optional RESTEasy
artifact.

Thus, I thought it might be better to keep them for now but, carefully
exclude the others to have only one API jar.

gsmet committed:
Add a proper JAXB implementation for quarkus-jaxb

The implementation is not present in JDK 11+ so we need it there. It was
hidden because of the IT having a dependency to an extension already
including it.

gsmet committed:
Be more creative about JAXB exclusion to avoid issues with REST Assured

Note that it's really just to avoid REST Assured dependency to be overly
verbose on JDK 11+ as the JAXB dependency is really needed there.

gsmet committed:
Adjust elytron-security-oauth2 IT build configuration

FroMage commented:
@gsmet `request.remoteAddress().host()` is how I've implemented it in `resteasy-vertx` as well, so that's fine.

Now, if you're talking about https://tools.ietf.org/html/rfc7239#section-4 and the previous headers for this, we should indeed have an answer. But the same question is true of the host/proto parts as well, because they should be used by `UriInfo`.

My guess is that this should be done at a lower-level, but perhaps I'm wrong.

FroMage commented:
I could not find anything related in netty. @vietj says vertx-core doesn't handle it either. I couldn't find anything in quarkus either. Perhaps we're missing handling for these headers entirely?

Do you know @stuartwdouglas ?

stuartwdouglas commented:
Undertow can handle it (io.undertow.server.handlers.ForwardedHandler), but I don't think vert.x really has any low level way of handling it. There is also X-Forwarded-For which is more common than the RFC one.

gsmet committed:
Fix dependency exclusions in Eclipse MicroProfile TCK

gsmet committed:
Also exclude javax.servlet:javax.servlet-api in addition to servlet-api

FroMage commented:
Yeah, I'm talking about the older headers as well as the new one. So it appears we just don't handle it at all in Quarkus, which is frankly surprising to me because I've always used forward proxies to deploy my apps. This means that we can't trust `UriInfo` to generate absolute links (port/proto/host name will be wrong) and we can't trust the remote IP either (I've had plenty of deployments that logged the remote IP on invalid logins so `fail2ban` would ban the IP after too many retries).

I will open an issue but I'm very surprised nobody noticed this before.

FroMage commented:
Added https://github.com/quarkusio/quarkus/issues/4922

gsmet committed:
Add a couple more javax exclusions + minor adjustments

gsmet committed:
Raise a warning if a banned dependency is used

For now, I only added the javax/jakarta/jboss ones but we could also add
the logging ones in the future.

gsmet committed:
Make a Curator message debug instead of info

kenfinnigan commented:
What is the likelihood of this being merged today?

I ask because I have a PR for MP REST Client TCKs that needs the new RESTEasy version

gsmet commented:
From my point of view: it's mergeable as soon as CI is green.

I just added some nice warnings for the users so there shouldn't be any migration issue.

But we need someone to review it and approve it. And people to generally agree with the approach.

kenfinnigan commented:
@gsmet FYI this PR now has a conflict and needs rebasing

gsmet commented:
This is going to be a lot of fun if it does not get merged quickly :).

gsmet commented:
@kenfinnigan rebased!

kenfinnigan commented on bom/runtime/pom.xml:
```diff
@@ -72,17 +80,15 @@
         <classmate.version>1.3.4</classmate.version>
-        <javax.el-impl.version>3.0.1-b11</javax.el-impl.version>
-        <hibernate-validator.version>6.1.0.Alpha6</hibernate-validator.version>
+        <jakarta.el-impl.version>3.0.3</jakarta.el-impl.version>
```
Should this be moved to where all the other `jakarta.*` properties are?

It can be done in a later PR if it's easier

gsmet replied:
Yeah, it used to be there because it was added with HV. But I can see how it makes sense to move them all at the same place now.

gsmet replied:
I reordered things.

kenfinnigan reviewed (approved):
It would be good if @stuartwdouglas can take a look as well.

Also, should the commits be squashed?

gsmet commented:
About the commits being squashed: I think we shouldn't squash them so that we can easily revert a part of the approach if it's deemed unpractical. There's a good chance we would keep some of them anyway.

kenfinnigan commented:
no worries, just asking

geoand commented:
Looks good to me as well, although my perspective is pretty limited

gsmet commented:
I tested the quickstarts and it was pretty much OK, except for the Kafka one which displayed a warning but this was an error in the quickstart which I fixed here: https://github.com/quarkusio/quarkus-quickstarts/pull/340/files .

stuartwdouglas commented:
This is a big change to make so close to 1.0. I am mostly ok with the implementation, although the banned dependency mechanism will need some work in the future (to make it pluggable, and also to allow banned dependencies to be automatically changed to good ones).

@emmanuelbernard @n1hility are you guys ok with a change this big so close to release?

gsmet commented:
Yeah, I agree the banned dependencies is very naive for now but it does the work for what we need currently.

I think with the dependencies banned by the enforcer and the warning mechanism, we are pretty safe. We will be able to tune things further if needed depending on the user reports.

emmanuelbernard commented:
Go.

Sanne commented:
awesome
